url5a01f2c1d9_core9_core_tgz-pJ8-io_core9_core_boot_BootstrapFrameworkJ8/src/net/xeoh/plugins/base/impl/classpath/cache/JARCache.java

```diff
@@ -0,0 +1,298 @@
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
+ * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * 
+ */
+package net.xeoh.plugins.base.impl.classpath.cache;
+
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.net.URI;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+/**
+ * Information about JAR files
+ * 
+ * @author rb
+ */
+public class JARCache {
+
+	/**
+	 * Information about a JAR plugin container.
+	 * 
+	 * @author Ralf Biedert
+	 * 
+	 */
+	public static class JARInformation implements Serializable {
+		/** */
+		private static final long serialVersionUID = 6734024814836912079L;
+
+		/** List of valid classes in the jar */
+		public List<String> classes = new ArrayList<String>();
+
+		/** List of valid plugins in the jar */
+		public Map<String, Collection<String>> subclasses = new HashMap<String, Collection<String>>();
+
+		/** */
+		public long lastAccess = System.currentTimeMillis();
+
+		/** */
+		public int usageCount = 0;
+
+		/** Is the contents information valid? */
+		public boolean classesValid = false;
+
+	}
+
+	/** */
+	final Logger logger = Logger.getLogger(this.getClass().getName());
+
+	/** Name of the default cache file */
+	private final static String DEFAULT_CACHE_FILE = "jspf.plugin.cache";
+
+	/** Is the cache enabled */
+	private boolean cacheEnabled = false;
+
+	/** maps a fingerprint to a jar information */
+	private Map<String, JARInformation> cacheMap = new HashMap<String, JARInformation>();
+
+	/** */
+	private boolean weakMode = false;
+
+	private String cachePath;
+
+	/**
+	 * Load cache
+	 */
+	@SuppressWarnings("unchecked")
+	public void loadCache() {
+		if (!this.cacheEnabled)
+			return;
+
+		final String cacheFile = (this.cachePath == null) ? DEFAULT_CACHE_FILE : this.cachePath;
+
+		try {
+			// lock the lockfile
+			final FileInputStream fis = new FileInputStream(cacheFile);
+			final ObjectInputStream ois = new ObjectInputStream(fis);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *ois*. The resource is referenced at lines 113, 114. The resource is closed at line 118. There are other execution paths that do not contain closure statements, for example, when ObjectInputStream.readObject throws an exception. Ensure *ois* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).